packages/svelte/src/internal/client/dev/ownership.js

```diff
@@ -98,60 +101,89 @@ export function mark_module_end(component) {
 }
 
 /**
- *
  * @param {any} object
  * @param {any} owner
+ * @param {boolean} [global]
  */
-export function add_owner(object, owner) {
-	untrack(() => {
-		add_owner_to_object(object, owner);
-	});
+export function add_owner(object, owner, global = false) {
+	if (object && !global) {
+		// @ts-expect-error
+		const component = current_component_context.function;
+		const metadata = object[STATE_SYMBOL];
+		if (metadata && !has_owner(metadata, component)) {
+			let original = get_owner(metadata);
+
+			if (owner.filename !== component.filename) {
+				let message = `${component.filename} passed a value to ${owner.filename} with \`bind:\`, but the value is owned by ${original.filename}. Consider creating a binding between ${original.filename} and ${component.filename}`;
+
+				// eslint-disable-next-line no-console
+				console.warn(message);
+			}
+		}
+	}
+
+	add_owner_to_object(object, owner);
 }
 
 /**
  * @param {any} object
  * @param {Function} owner
  */
 function add_owner_to_object(object, owner) {
-	if (object?.[STATE_SYMBOL]?.o && !object[STATE_SYMBOL].o.has(owner)) {
-		object[STATE_SYMBOL].o.add(owner);
-
-		for (const key in object) {
-			add_owner_to_object(object[key], owner);
+	const metadata = /** @type {import('#client').ProxyMetadata} */ (object?.[STATE_SYMBOL]);
+
+	if (metadata) {
+		// this is a state proxy, add owner directly
+		(metadata.owners ??= new Set()).add(owner);
```

Comment:
Pretty sure this is wrong, because `null` means "everything's allowed", [causing unwanted narrowing in a case like this](https://svelte-5-preview-gu560ep2h-svelte.vercel.app/#H4sIAAAAAAAACo2Q0WrEIBBFf2UYCpuwIemzTQJLPqPpgzEu2CYqOllaxH8vuu1uYKH0zXGuc5wT8KwW6ZG9BtR8lcjwZC1WSF82Ff4iF5JYoTebE-mm9cIpS_2oR1KrNY4ggCdOEiKcnVnhUDe8vj48vOxiw6097Nptcx-o2wEmpWfGu5BH1he-bDJC04-6nTYio8FosSjx0YWihK6HXa4WZtN0PMb8uZzyDMJDIibqdVqPFa5mVmclZ2TkNhmrm4nfLe463v1ehfzMewmjPf0o6OApH4qQeQxCJjJ4jrH8izX81_kik3CeSEkVnxZZlBBTbZ2xvij3Uh-Rb_EbH6ggnfQBAAA=). Looking into it.

Reply:
Pushed a fix